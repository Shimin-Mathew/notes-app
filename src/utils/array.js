export default function lastOf(array) {
	if (typeof array.length !== 'number')
		return null;

	return array[array.length - 1];
}