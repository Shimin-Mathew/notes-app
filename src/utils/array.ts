/**
 * Return the last item of the Arraylike
 * @param array - the array
 * @returns lastItems
 **/
export default function lastOf<T = any>(array: T[]) {
	if (array.length === 0) return null;

	return array[array.length - 1];
}
