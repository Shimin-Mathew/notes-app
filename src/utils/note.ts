export function getTitleFromNote(note: string) {
	if (!note) return;

	const firstFewWords = note.split(' ').slice(0, 4).join(' ');
	const firstWordsSliced = firstFewWords.length > 12
		? firstFewWords.slice(0, 12)
		: firstFewWords;
	return `${firstWordsSliced}...`;
}