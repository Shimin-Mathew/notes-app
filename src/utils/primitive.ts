export function asNumber(number?: string | number, fallback = 0) {
	if (number === undefined) return NaN;

	const _num = typeof number === 'number' ? number : parseInt(number);
	return typeof _num === 'number'? _num: fallback;
}

export function asString(value: any) {
	return typeof value === 'string' ? value : String(value);
}

export function joinString(strings: (string | undefined)[], separator = ' ') {
	return strings.filter(i => i).join(separator);
}