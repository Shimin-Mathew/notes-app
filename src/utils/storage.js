const storeKey = 'todos';
const fqCountKey = 'visits';

const get = key => localStorage.getItem(key);
const set = (key, value) => localStorage.setItem(key, value);

// NOTES
export function loadTodos() {
	return JSON.parse(get(storeKey)) || [];
}

export function persistTodos(todos) {
	set(storeKey, JSON.stringify(todos));
}

// VISIT COUNTERS
function getFq(fallback = {}) {
	return JSON.parse(get(fqCountKey)) || fallback;
}

function setFq(value) {
	set(fqCountKey, JSON.stringify(value));
}

export function initVisitCounters() {
	let newCount = {};
	const vistFq = getFq();
	const persistedTodos = loadTodos();

	persistedTodos.forEach(todo => {
		newCount[todo.id] = vistFq[todo.id] || 1;
	});

	setFq(newCount);
}

export function increaseCount(noteId) {
	const vistFq = getFq();
	vistFq[noteId] += 1;
	setFq(vistFq);
}

export function getMostVisited() {
	let maxId = null;
	let maxValue = 0;
	const vistFq = getFq();

	if (vistFq !== null) {
		Object.keys(vistFq).forEach(id => {
			if (maxValue < vistFq[id]) {
				maxValue = vistFq[id];
				maxId = id;
			}
		});
	}

	return maxId;
}
