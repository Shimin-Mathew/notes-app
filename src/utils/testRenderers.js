import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';

import rootReducer from 'reducers';
import todos from 'components/todoList/__test__/todos';
import { Router } from 'react-router';
import { createMemoryHistory } from 'history';

const dummyStore = createStore(rootReducer, {todos});
const dummyHistory = createMemoryHistory();

export function renderWithStore(component) {
	return render(
		<Provider store={dummyStore}>{component}</Provider>
	);
}

export function renderWithRouter(component) {
	const wrap = render(<Router history={dummyHistory}>{component}</Router>);
	const rerender = component => wrap.rerender(<Router history={dummyHistory}>{component}</Router>)
	return {...wrap, history: dummyHistory, rerender}
}

export default function renderWithStoreAndRouter(component) {
	const wrap = render(<Router history={dummyHistory}>{component}</Router>);
	return {...wrap, history: dummyHistory}
}