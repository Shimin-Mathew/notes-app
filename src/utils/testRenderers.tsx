import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';

import rootReducer from 'reducers';
import notes from 'components/noteList/__test__/dummyNotesData';
import { Router } from 'react-router';
import { createMemoryHistory, Location } from 'history';
import { match, RouteComponentProps } from 'react-router-dom';

const dummyStore = createStore(rootReducer, { notes });
const dummyHistory = createMemoryHistory();

export function renderWithStore(component: React.ReactElement) {
	return render(<Provider store={dummyStore}>{component}</Provider>);
}

export function renderWithRouter(component: React.ReactElement) {
	const wrap = render(<Router history={dummyHistory}>{component}</Router>);
	const rerender = (component: React.ReactElement) =>
		wrap.rerender(<Router history={dummyHistory}>{component}</Router>);
	return { ...wrap, history: dummyHistory, rerender };
}

export default function renderWithStoreAndRouter(component: React.ReactElement) {
	const wrap = render(<Router history={dummyHistory}>{component}</Router>);
	return { ...wrap, history: dummyHistory };
}
// FIXME:  doesn't return match<T>
export function createMatch<T extends {}>(options: Partial<match<T>> = {}) {
	const match: match = {
		url: '',
		path: '',
		isExact: true,
		params: {},
		...options
	};
	return match;
}

export function createLocation<State extends {}>(options: Partial<Location<State>> = {}) {
	const location: Location = { hash: '', key: '', pathname: '', search: '', state: null, ...options };
	return location;
}

export function createRouterComponentProps(): RouteComponentProps {
	return {
		match: createMatch(),
		location: createLocation(),
		history: dummyHistory
	}
}