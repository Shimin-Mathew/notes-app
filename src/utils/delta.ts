import { getTextFromTodo, isDoneTodo, isTodo } from 'utils/todos';

type TextBlock = { type: 'text'; data: string; };

type TodoItem = { text: string; isDone: boolean };

type TodoBlock = { type: 'todo'; data: TodoItem[]; };

type HeadingBlock = { type: 'heading', data: { headingText: string, level: number }};

type ParsedData = { currentTodoList?: any[]; data: DeltaData[]; };

type DeltaData = TextBlock | TodoBlock | HeadingBlock;

export function getDeltaFromText(text: string) {
	const lines = text.split('\n');
	const initialParsedData: ParsedData = { data: [] };

	const parsedData = lines.reduce((acc, line) => {
		if (isTodo(line)) {
			const todoItem = {
				text: getTextFromTodo(line),
				isDone: isDoneTodo(line)
			};
			if (Array.isArray(acc.currentTodoList)) {
				acc.currentTodoList.push(todoItem);
			} else {
				const newtodoList = [todoItem];
				acc.data.push({
					type: 'todo',
					data: newtodoList
				});
				acc.currentTodoList = newtodoList;
			}
		} else if (line.startsWith('#')) {
			let level = 0;
			let headingText = '';
			let isBroken = false;
			line.split('').forEach(ctr => {
				if (!isBroken) {
					if (ctr === '#') {
						++level;
						return;
					} else {
						isBroken = true;
					}
				}
				headingText += ctr;
			});
			acc.data.push({
				type: 'heading',
				data: { headingText: headingText, level: level>6? 6: level }
			});
		} else {
			acc.currentTodoList = undefined;
			acc.data.push({
				type: 'text',
				data: line
			});
		}
		return acc;
	}, initialParsedData);

	parsedData.currentTodoList = undefined;

	return parsedData.data;
}

export function getTextFromDelta(delta: DeltaData[]) {
	if (getIsEmptyDelta(delta)) {
		return delta[0].data as string;
	}

	const deltaAsString = delta.reduce((acc, lineData, lineIndex) => {
		const lineSeparator = lineIndex > 0? '\n': '';
		if (lineData.type === 'todo') {
			const todolistAsString = lineData.data.reduce(
				(innerAcc, todo, todoIndex) =>
					todo
						? innerAcc + (todoIndex > 0? '\n': '') + `${todo.isDone ? '*' : 'o'} ${todo.text}`
						: innerAcc,
				''
			);
			acc += lineSeparator + todolistAsString;
		} else if (lineData.type === 'heading') {
			acc += lineSeparator + '#'.repeat(lineData.data.level) + lineData.data.headingText;
		} else {
			acc += lineSeparator + lineData.data;
		}

		return acc;
	}, '');

	return deltaAsString
}

export function getIsEmptyDelta(delta: DeltaData[]) {
	return delta.length === 1 && delta[0].type === 'text' && delta[0].data === '';
}