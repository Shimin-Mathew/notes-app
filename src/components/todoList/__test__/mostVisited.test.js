import React from 'react';
import MostVisited from '../mostVisited';
import {renderWithStore} from 'utils/testRenderers';

jest.mock('utils/storage');

test('mostvisited', () => {
	const { asFragment } = renderWithStore(<MostVisited />);
	expect(asFragment()).toMatchSnapshot();
})