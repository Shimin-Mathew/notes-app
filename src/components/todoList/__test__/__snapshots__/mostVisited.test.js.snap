// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mostvisited 1`] = `
<DocumentFragment>
  <article>
    <div>
      <button
        class="todo-controls"
      >
        Edit
      </button>
      <select
        class="todo-controls"
      >
        <option
          value="high"
        >
          High
        </option>
        <option
          value="med"
        >
          Medium
        </option>
        <option
          value="low"
        >
          Low
        </option>
      </select>
      <button
        class="todo-controls"
      >
        Delete
      </button>
    </div>
    <label>
      Title:
      <input
        class="edit-note-title"
        disabled=""
        name="title"
        placeholder="What do you need to do ?"
        value="app"
      />
    </label>
    <label>
      Note Content:
      <textarea
        class="edit-note-text"
        disabled=""
        name="notes"
        placeholder="Write notes on:
 * How to achieve the goal
 * Deadlines
 * Who is going to help ?"
      >
        verthe
      </textarea>
    </label>
  </article>
</DocumentFragment>
`;
